include/itkMaxPhaseCorrelationOptimizer.hxx

```diff
@@ -82,7 +77,125 @@ MaxPhaseCorrelationOptimizer< TRegistrationMethod >
+      for (; !oIt.IsAtEnd(); ++iIt, ++oIt)
+        {
+        typename ImageType::IndexType ind = oIt.GetIndex();
+        IndexValueType dist = 0;
+        for (unsigned d = 0; d < ImageDimension; d++)
+          {
+          IndexValueType distDirect = ( directExpectedIndex[d] - ind[d] ) * ( directExpectedIndex[d] - ind[d] );
+          IndexValueType distMirror = ( mirrorExpectedIndex[d] - ind[d] ) * ( mirrorExpectedIndex[d] - ind[d] );
+          if ( distDirect <= distMirror )
+            {
+            dist += distDirect;
+            }
+          else
+            {
+            dist += distMirror;
+            }
+          }
+
+        typename ImageType::PixelType pixel = 1000*iIt.Get() * std::exp( distancePenaltyFactor * dist ); //TODO: remove 1000 factor
+        oIt.Set( pixel );
+        }
+    },
+    nullptr );
+
+  WriteDebug( iAdjusted.GetPointer(), "iAdjusted.nrrd" );
+
+  // supress trivial zero solution
+  FixedArray< double, ImageDimension > dimFactor; // each dimension might have different size
+  for ( unsigned d = 0; d < ImageDimension; d++ )
+    {
+    double dimFactor = 100.0 / size[d]; // turn absolute size into percentages
+    }
+  constexpr IndexValueType znSize = 4; // zero neighborhood size
+  mt->ParallelizeImageRegion<ImageDimension>( lpr,
+    [&]( const typename ImageType::RegionType& region )
+    {
+      ImageRegionIteratorWithIndex< ImageType > oIt(iAdjusted, region);
+      for (; !oIt.IsAtEnd(); ++oIt)
+        {
+        bool pixelValid = false;
+        typename ImageType::PixelType pixel;
+        typename ImageType::IndexType ind = oIt.GetIndex();
+        IndexValueType dist = 0;
+        for ( unsigned d = 0; d < ImageDimension; d++ )
+          {
+          dist += ind[d] - oIndex[d];
+          }
+        if ( dist < znSize ) // neighborhood of [0,0,...,0]
+          {
+          pixel = oIt.Get();
+          pixel *= ( dist + 3 ) / ( m_ZeroSuppression + dist + 3 );
+          pixelValid = true;
+          }
+
+        for ( unsigned d = 0; d < ImageDimension; d++ ) // lines/sheets of zero indices
+          {
+          if ( ind[d] == oIndex[d] ) // one of the indices is "zero"
+            {
+            if ( !pixelValid )
+              {
+              pixel = oIt.Get();
+              pixelValid = true;
+              }
+            IndexValueType distD = ind[d] - oIndex[d];
+            if ( distD > IndexValueType( size[d] / 2 ) ) // wrap around
+              {
+              distD = size[d] - distD;
+              }
+            double distF = distD * dimFactor[d];
+            // avoid the initial steep drop of 1/(1+x) by shifting it by 3% of image size
+            pixel *= ( distF + 3 ) / ( m_ZeroSuppression + distF + 3 );
+            }
+          }
+
+        if ( pixelValid ) // either neighborhood or lines/sheets has updated the pixel
+          {
+          oIt.Set( pixel );
+          }
+        }
+    },
+    nullptr );
+
```

Comment:
Debug

Reply:
WriteDebug later gets moved from here to `itkPhaseCorrelationOptimizer.hxx`.